lib/cloudex/cloudinary_api.ex

```diff
@@ -40,6 +40,105 @@ defmodule Cloudex.CloudinaryApi do
+      {:ok, raw_response} ->
+        send_telemetry_event(start_time, :upload_large, size, raw_response)
+        response = @json_library.decode!(raw_response.body)
+        handle_response(response, item)
+      {:error, raw_response} ->
+        send_telemetry_event(start_time, :upload_large, size, raw_response)
+        error_response = @json_library.decode!(raw_response.body)
+        handle_response(%{"error" => %{"message" => error_response}}, item)
+      _ ->
+        send_telemetry_event(start_time, :upload_large, size)
+        handle_response(%{"error" => %{"message" => "Error uploading file"}}, item)
+    end
+  end
+
+  defp upload_chunks(item, chunk_size, size, unique_id, opts) do
+    File.stream!(item, [], chunk_size)
+    |> Stream.with_index()
+    |> Enum.reduce_while({:ok, %{}}, fn chunk_with_index, _acc ->
+      start_time = System.monotonic_time(:millisecond)
+      chunk_resp = upload_chunk(chunk_with_index, item, size, chunk_size, unique_id, opts)
+      case chunk_resp do
+        {:ok, %HTTPoison.Response{status_code: 200} = resp} ->
+          send_telemetry_event(start_time, :upload_chunk, chunk_size, resp)
+          {:cont, {:ok, resp}}
+        {:ok, resp} ->
+          send_telemetry_event(start_time, :upload_chunk, chunk_size, resp)
+          {:halt, {:error, resp}}
+        {:error, error} ->
+          send_telemetry_event(start_time, :upload_chunk, chunk_size)
+          {:halt, error}
+      end
+    end)
+  end
+
+  defp upload_chunk(
+         {chunk, index} = _chunk_with_index,
+         _file_path,
+         size,
+         chunk_size,
+         unique_id,
+         opts
+       ) do
+    start_byte = index * chunk_size
+
+    end_byte =
+      if div(size, chunk_size) == index, do: size - 1, else: start_byte + chunk_size - 1
+
+    content_range = "bytes #{start_byte}-#{end_byte}/#{size}"
+
+    chunk_headers = [
+      {"X-Unique-Upload-Id", unique_id},
+      {"Content-Range", content_range},
+      {"Content-Type", "multipart/form-data"},
+    ]
+
+    options = prepare_signed_opts(opts)
+
+    form = {:multipart,
+         [
+           {"file", chunk, {"form-data", [{"name", "file"}, {"filename", "blob"}]},
+            [{"content-type", "application/octet-stream"}]},
+           {"api_key", Cloudex.Settings.get(:api_key)} | options
+        ]}
+
+    url = "#{@base_url}#{Cloudex.Settings.get(:cloud_name)}/video/upload"
+    request_options = opts[:request_options] || []
+
+    HTTPoison.post(
+      url,
+      form,
+      chunk_headers,
+      credentials() ++ request_options
+    )
+  end
+
+  defp send_telemetry_event(start_time, function, size, resp) do
+    end_time = System.monotonic_time(:millisecond)
+    resp_result = if resp.status_code == 200, do: :success , else: :failure 
+    :telemetry.execute([:cloudex, function, resp_result],
+      %{request_time: end_time - start_time},
+      %{status_code: resp.status_code, bytes: size}
```

Comment:
it would probably be good to surface the content-range headers or pass down some arguments to keep track of how many bytes are done and how many are left, maybe the original filename and stuff like that too